### ISSUE TITLE
Support for dm-crypt/LKS key drop/readd on suspend to ram

### ISSUE THREAD

**Issue by jo-so** [systemd/systemd#5794]
### Submission type

  - [ ] Bug report
  - [X] Request for enhancement (RFE)

### Request

The kernel supports the temporary removal of the encryption keys of luks devices, which should be done before the suspend to RAM gets triggered. After the wake-up the key can be feed back to the kernel and all operations can continue. See [luksSuspend and luksResume](http://man7.org/linux/man-pages/man8/cryptsetup.8.html). Before systemd issues the suspend command to the kernel, it should drop the keys of encrypted disks and after wake-up prompt the user for the password to reopen the encrypted device.

This should support encrypted root devices, i.e. there can be no access to the disk (e.g. logging) after luksSuspend and before luksResume.

### Idea

Add this support to systemd-sleep. Either by temporary creating a chroot with tmpfs and trigger the suspend with a process running there or using mlock(2) to pin the process' memory.

**Comment by poettering** [comment]
Uh, oh, making block devices unavailable while most userspace running off it stays around is highly problematic, and I am not convinced this is really strong enough a usecase to add this immense complexity...

**Comment by rokups** [comment]
A workaround in this case is chrooting into initramfs. Example can be seen [here](https://github.com/vianney/arch-luks-suspend). This would be a great addition to systemd as aside from hacks from community nothing i know of supports suspending encrypted disks and that makes suspend-to-ram a no-go for encrypted systems.

**Comment by JeWe37** [comment]
Apparently about a year or two ago there [was some work on this by some debian devs](https://blog.freesources.org/posts/2020/08/cryptsetup-suspend/) and cryptsetup on debian now has support for this after [this fork](https://salsa.debian.org/mejo/cryptsetup-suspend) was merged. There doesn't seem to have been any work it since though despite some issues remaining unresolved.

The authors discovered that indeed as @poettering already guessed there were some complications when attempting to suspend and resume the devices beyond what the existing libraries built for this sort of suspension do. In the debian implementation they additionally look to freeze all processes before they suspend the devices, but still ran into some issues with for instance potentially running out of memory due to argon2e's high memory requirements.

I'm not sure but I think this feature could potentially benefit from tighter integration with systemd. Properly freezing the processes should be easier directly from systemd. Mainly you would need an all-process cgroup or something similar to freeze everything properly.

Especially now that better support for fully handling crypto devices has been merged into systemd I think this might be worth another look. I might also look into it at some point should I find the time.
